src/Tools/dotnet-monitor/CollectionRules/Actions/CollectDumpAction.cs

```diff
@@ -3,19 +3,76 @@
 // See the LICENSE file in the project root for more information.
 
 using Microsoft.Diagnostics.Monitoring.WebApi;
+using Microsoft.Diagnostics.Monitoring.WebApi.Models;
+using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Exceptions;
 using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
+using Microsoft.Extensions.DependencyInjection;
 using System;
+using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
+using Utils = Microsoft.Diagnostics.Monitoring.WebApi.Utilities;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
 {
-    internal sealed class CollectDumpAction :
-        ICollectionRuleAction<CollectDumpOptions>
+    internal sealed class CollectDumpAction : ICollectionRuleAction<CollectDumpOptions>
     {
-        public Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        private readonly IDumpService _dumpService;
+        private readonly IServiceProvider _serviceProvider;
+
+        public CollectDumpAction(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            _dumpService = serviceProvider.GetRequiredService<IDumpService>();
+        }
+
+        public async Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
         {
-            throw new NotImplementedException("TODO: Implement action");
+            DumpType dumpType = options.Type.GetValueOrDefault(CollectDumpOptionsDefaults.Type);
+            string egressProvider = options.Egress;
+
+            string dumpFileName = Utils.GenerateDumpFileName();
+
+            string dumpFilePath = string.Empty;
+
+            // Given our options validation, I believe this is probably redundant...should I remove it?
+            if (string.IsNullOrEmpty(egressProvider))
```

Comment:
Should this stay or go?

Reply:
Rather than checking individual fields, validate the object as it was designed to be validated:
```cs
ValidationContext context = new(options, _serviceProvider, items: null);
Validator.ValidateObject(options, context, validateAllProperties: true);
```
This will throw an exception if any of the properties are invalid.